### ISSUE TITLE
Controlling which color spaces appear in menus

### ISSUE THREAD

**Issue by doug-walker** [AcademySoftwareFoundation/OpenColorIO#1222]
There has been wide agreement that it would be helpful for config authors to be able to limit which color spaces appear in application menus.  Indeed, several features of OCIO v2 are aimed at improving this situation and a number of PRs have been merged.  At this point, it seems that it would be useful to summarize this topic and pull together the results of various conversations to make sure that we are all on the same page as to how this will work in v2 and decide on any remaining tweaks.

NOTE:  To clarify the scope of this issue, it is only concerning color space menus.  Applications also need to create menus for displays and views, but the display/view section of the config already filters a list of views based on what display is selected.  Furthermore, the new viewing rules allows an application to further filter the views based on the color space of the image being viewed.  So display/view filtering is out of scope for the present issue.

## OVERVIEW

To recap and frame the color space menu problem, there are several new and existing mechanisms that are relevant:

1. The roles may be used to specify that there should be no user choice (i.e., a fixed color space should be used).  An example of this that is used in Maya is the "rendering" role.  If this role is present, the application rendering space menu in Preferences is locked to this color space.

2. The new inactive_colorspaces property of a config or the corresponding environment variable may be used to remove color spaces from application menus while still allowing them to be used, if necessary.  For example, an old scene file may use deprecated color space names.  Leaving them in the config allows an application to read and process the old files but adding those color spaces to the inactive list means they won't show up in menus by default.  Note that an application may temporarily add a given space to the menu in this scenario when displaying what was assigned to a given asset in the scene.

3. The proposed "aliases" feature (Issue #1129) would provide another way to prevent color spaces from appearing in menus while still allowing them to be used elsewhere (e.g., to be used in file rule paths).

4. An application may include all roles in color space menus in special circumstances.  This statement might seem contrary to several discussions in TSC meetings concluding that it is generally undesirable for applications to add roles to color space menus.  (If a config author wants them to be in the menu, a color space with the desired name could be added.)  However, there are some situations where roles could be appropriate.  For example, if an application allowed editing of file rules, it would be appropriate to allow a role to be selected as a rule's color space.

5. The family attribute of a color space may be used to group the menu contents into hierarchical menus.  (This does not actually control whether a space appears but could be used to similar effect, e.g., by putting spaces under a "DO NOT USE" heading, which is sometimes used as a workaround in v1.)

6. Other color space attributes such as encoding, isdata, and equalitygroup could also be used to filter color spaces.

7. The new display_colorspaces section of the config could provide another means for an applications to sort types of color spaces.

8. And last, but not least, the primary purpose of the new categories attribute of a color space is to filter what shows up in each menu.

## APP HELPERS

In order for these capabilities to be implemented in a uniform manner, there needs to be some example code provided.  The ColorSpaceMenuHelper class in the App Helpers code is intended to serve this purpose and already supports a number of the above mechanisms (nos. 1, 2, 4, 5, and 8).  The App Helpers were introduced in PR #952 and PR #1220 (in progress) moves it into the core library.

## CATEGORIES

The item that needs the most elaboration is the last one: categories.  The categories feature was added in PR #627 and has been discussed in Issue #623 and at many meetings.  However we still need to get alignment on what set of official strings will be documented and how the App Helpers code should filter based on them.

During the discussions, there have been two main use-cases that have been discussed:

Menu-based (or Task-based) filtering -- In this scenario, the categories contain a description of the task that a given application menu is used for.  For example, Maya has a menu for assigning a color space to a file texture and another menu for selecting the rendering (working) space.  Clearly, a different set of color spaces are appropriate for those two tasks.  From an app developer's point of view, these category string(s) would be predetermined and fixed for a given menu (and they should be documented by the app developer for the benefit of config authors).  A baseline set of strings should be documented by OCIO to promote consistency among developers but, as with roles, it should be possible to introduce more without requiring any library changes.

Artist-based filtering -- In this scenario, the categories contain a description of the type of person using the application based on what type of tasks they focus on.  For example, in a large studio there are departments for modeling, layout, look-dev, lighting, compositing, and many others.  The degree to which a given sort of artist is knowledgeable about color spaces and is expected to be able to select which ones are appropriate varies a lot with their role.  Hence it makes sense to filter color spaces based on this.  From an app developer's point of view, the category string(s) being looked for are *not* predetermined and the expectation is that they would need to be set at runtime (e.g., by an environment variable).  However, it would still be useful for OCIO to document a baseline set of strings to promote consistency (and again with the understanding that it does not need to be an exhaustive list).

Another way of looking at this is that the app developer hard-codes one set of keys (the menu categories) and the app end-user provides another set of keys at runtime (the artist categories) and the intersection of these is what filters what is provided by the config author.  In other words, another way of framing the problem is to think of it in terms of a set of hard-coded app-based categories plus a set of runtime end-user-based categories.

That brings us to the question of how the App Helpers example code should filter color space lists using categories.  Currently, the code takes one list of strings and will select all color spaces that contain any of those strings (in other words, a union or logical OR).  If no strings are provided, or the config does not use categories, the color spaces are not filtered by them.

The limitation of the existing algorithm is that it does not allow expressing a logical AND relationship, which seems like it would be necessary to capture the menu-type AND artist-type intersection that is desired.

This issue proposes extending the App Helpers code to take two lists of categories rather than one -- an app-supplied list and an end-user-supplied list.  The following algorithm would then be used:

```
if (no categories are provided to App Helpers)  --> don't filter based on categories

app_set = FindColorSpaces(app_categories)
end_user_set = FindColorSpaces(end_user_categories)

if ( !isempty(app_set) && !isempty(end_user_set) )  --> use the intersection of the two sets
else if ( !isempty(app_set) )  --> filter using the app categories
else if ( !isempty(end_user_set) )  --> filter using the end-user categories
else  --> don't filter based on categories
```

## MENU-BASED CATEGORIES STRINGS

Here is a proposed baseline set of Menu-based categories:

- file-io -- For menus setting the color space of imported or exported files.
- working-space -- For menus setting a working color space.
- look-process-space -- For menus setting the process space for applying a CDL or LUT. 

## OTHER FILTERING

Note that the above menu-based categories are intended to be used in connection with the "encoding" property of the color space (Issue #1031, merged in PR #1038).  So for example, rather than define "scene_linear_working_space", "log_working_space", etc., there is only a "working-space" category.  If an application needs scene-linear working spaces, they would provide App Helpers with a category of "working-space" and an encoding of "scene-linear".  Hence another refinement proposed here is to allow an encoding to be presented to the ColorSpaceMenuHelper class (which is not currently implemented).

Similarly, it seems that it could be useful to further filter the list to show either all color spaces, only display color spaces (Issue #926, merged in PR #939), or only non-display color spaces.  (The Config class API already supports this, but it could be useful to add that to ColorSpaceMenuHelper too.)

## ARTIST-BASED CATEGORY STRINGS

The Artist-based categories are a bit harder to enumerate.  There are a lot of different department names in a big VFX or animation studio and it might be difficult to agree on a representative list.  Furthermore, there may be little point to distinguish between certain types of artists if they have similar needs in terms of color management (e.g., modeling vs. layout).  So perhaps one way of approaching it is to delineate between 3d (modeling, animation, lighting, etc.) and 2d (editorial, compositing, etc.) departments/jobs and based on the amount of access needed?  Here is a proposed baseline set:

- basic-3d -- For artists in 3d departments needing only basic color management.
- advanced-3d -- For artists in 3d departments needing advanced color management access.
- basic-2d -- For artists in 2d departments needing only basic color management.
- advanced-2d -- For artists in 2d departments needing advanced color management access.

## SUMMARY

In summary, here is the proposal:

Add to ColorSpaceMenuHelper in App Helpers:
- The ability to specify a second list of categories, to be used in intersection with the first (existing) set (thereby allowing both an app-based set and an end-user-based set).
- The ability to specify an encoding to further filter the categories results.
- The ability to further filter the results to all color spaces, only display color spaces, or only non-display color spaces.

Document the set of Menu-based and Artist-based categories listed above as the recommended baseline set.

**Comment by zachlewis** [comment]
_Dig it!_

Here are some thoughts and ideas, increasingly out-of-scope:

- I think the proposed algorithm for handling app-demanded and user-provided categories is great. 

- I also really like the idea of using the color space "encoding" property to get menu-based categories "for free." I suppose "encoding" values are expected (but not required?) to be one of: [`data`, `scene-linear`, `display-linear`, `log`, `sdr-video`, `hdr-video`, `sdr-cinema`, `hdr-cinema`] ...? 

- Any consideration given to sort order? How are merged / filtered color spaces presented? Are hierarchical menus preserved when presenting filtered color spaces? Should special consideration be given to colorspaces that are assigned to certain roles? (e.g., floating the color space assigned to the "color_timing" role to the top of a "look-process-space" menu?) 

- I _would_ like the option to enable a standard "strict-sorting" method that forces families, followed by color spaces, to be listed in alphabetical order. (And I think it should be an option specified in the config, as opposed to an environment variable set at runtime). 

- Even though we have other ways of explicitly enabling / disabling color space visibility, I still think users would find it useful to be able to hide (from unfiltered menus) color spaces belonging to certain families. I was thinking an intuitive mechanism for this might be to filter any color spaces whose family string is prefixed with a period, or includes a separator delimiter immediately followed by a period. For instances, color spaces under the  ".DO_NOT_USE" and "foo/.bar" families would be hidden, but not necessarily color spaces directly under "foo" or "foo/mew" or what have you. Thoughts? 

- How are ViewTransform categories used?

**Comment by Shootfast** [comment]
I think this looks great. The only concern I have is regarding the pre-defined menu-based category strings (
    file-io, working-space, look-process-space, etc). 
I think we should take the lessons learnt from the existing default "roles" list, that has proven to be a source of confusion to both DCC authors and end-users alike.
There will be different opinions on whether a particular DCC's drop-down represents "file-io" or "stream-io" or any other unique concept, and so the best way to avoid confusion would be for DCC's to empower users to tag each menu with their own appropriate category, effectively letting them choose to implement the menu or artist role style of categorization as they see fit. This would be a little outside the scope of OCIO itself, but we can steer the app-helpers code towards providing this as the desired workflow.

**Comment by KevinJW** [comment]
As mentioned in the TSC meeting, I'd prefer something explicit over 'magic' hiding based on naming of the family string pattern matching some set of values.

Kevin

**Comment by doug-walker** [comment]
This proposal has now been implemented in PR #1220.  Thanks for the comments above, here are the responses based on the latest work:

Zach, 
- Yes you are correct about the encodings.  
- The sort order basically preserves the ordering of the config, please refer to the comments in OpenColorAppHelpers.h for more details.  
- Hierarchical menus are preserved by the filtering, but note that the application ultimately controls how they order and present the hierarchical menus since creating them is more complicated than simply iterating over a list.  
- The ViewTransform categories are not currently used but the idea was that they could ultimately be used to provide some additional filtering on Views, similar to what has been done for color spaces.

Mark, per your request, we have not added the proposed category strings listed above into the header files.  Currently they are only used in the unit tests as plausible real-world examples.

**Comment by doug-walker** [comment]
Closing this as fixed via PR #1220.
